message/messaging_service.cc

```diff
@@ -625,7 +625,12 @@ shared_ptr<messaging_service::rpc_protocol_client_wrapper> messaging_service::ge
         remove_error_rpc_client(verb, id);
     }
 
-    auto must_encrypt = [&id, &verb, this] {
+    auto addr = get_preferred_ip(id.addr);
+    auto broadcast_address = utils::fb_utilities::get_broadcast_address();
+    bool listen_to_bc = _cfg.listen_on_broadcast_address && _cfg.ip != broadcast_address;
+    auto laddr = socket_address(listen_to_bc ? broadcast_address : _cfg.ip, 0);
```

Comment:
What is the minimal information we need to decide if we should use encryption between two nodes?
I think what really matter is the dc and rack of the two nodes, no matter what the listen address or broadcast address a node is using. 

If this is true, I guess we can build a map

std::unordered_amp<gms::inet_address,  bool> should_encrypt_map_for_peer_nodes;

to answer if we should use encrypt or not for a given ip address. 

We put both the internal and broadcast ip for a remove node inside the map.

This way, there is no need to run a lot of complicated check below.

Reply:
The requirement for downgrading to non-encrypted is (depending on setting) that we are in either same rack or dc - _using/based on the ip addresses selected for the connection_. We have mappings for "should encrypt" (in a sense) - it is the rack/dc association for endpoints. The problem is again when we use addresses that are _not_ the endpoint address, but the "internal" ones. 

The code here makes us look more strictly at the actual IP:s used for connecting/binding to, not the "official" endpoint addresses (i.e. our broadcast and the argument `endpoint` - the latter of which is replaced with a "preferred" address). 

As stated above, in the current world, without #9714, this will effectively force full TLS for all connections via internal network on ec2 etc. But at least it will not fail.

Reply:
To clarify: You cannot solve the problem by simply mapping IP:s to encrypt or don't encrypt, since it is determined by the tuple source, dest and encryption config. You could compute it once, per endpoint and node (assuming we don't change config while live), but it would pretty much just be the above calculation. The root problem is still not having endpoint association for the internal network addresses, and thus not being able to determine where a node belongs.

Reply:
Can't we get the answer if we should encrypt or not if we know the destination ip address and encryption config? Why does what the local node listens to matter when we make the encrypt decision. Connecting to a remote node using the public ip address or private ip address are not supposed to have impact on the encryption decision.

For example, say we have internode_encryption = rack, and node 1 an node 2 in the same dc but different racks, no matter node 1 talks to node 2 using node 2's public or internal ip address, we should always encrypt because they are in different racks.  So we have two entries on node1 for node2:

encrypt_decision_map {
    {node2_internal_ip, yes}
    {node2_public_ip, yes} 
}

No matter how messages playing tricks to preferred ip, it does not have impact on the decision. We always get the correct answer.

Reply:
The problem is not knowing whether to encrypt or not. The old code selected "correctly", if we only look at the config. The problem is identifying an incoming connection on the other side (not this code, but the connection filter in the message server listening part). The incoming connection will typically be on a private network interface, coming from private nwif on the other side, so when that code checks these two IP:s against the rack/dc conditions, it cannot validate it -> dropped connection. 

The only thing we do here is making the check _stricter_ so we _don't_ encrypt some things we technically should encrypt, unless our gossip/snitch info can give us equivalent info using the actual private addresses we end up using. Thus, without #9714, we will for most ec2 cases _not_ encrypt now, _but_ we will otoh not fail on the other end either. 

With #9714, we will have strict check here + working tls downgrade.

Reply:
Let's add concrete examples for what is really going on. So you are saying the problem is that

Assume the cluster uses encrypt_what::rack, we have

node1: public1, private1
node2: public2, private2, 7000 (non ssl), 70001 (ssl)

when node1 uses private1 to talk to node2 private2:7000, since node2 can not get rack or dc info using private1, so filter_connection at line 269 will return false, then the connection will be rejected? Anything else can go wrong? If we have this problem, how does encryption ever work?

```
 254     if (!_server[0] && _cfg.encrypt != encrypt_what::all && _cfg.port) {
 255         auto listen = [&] (const gms::inet_address& a, rpc::streaming_domain_type sdomain) {
 256             so.streaming_domain = sdomain;
 257             so.filter_connection = {};
 258             switch (_cfg.encrypt) {
 259                 default:
 260                 case encrypt_what::none:
 261                     break;
 262                 case encrypt_what::dc:
 263                     so.filter_connection = [](const seastar::socket_address& addr) {
 264                         auto& snitch = locator::i_endpoint_snitch::get_local_snitch_ptr();
 265                         return snitch->get_datacenter(addr) == snitch->get_datacenter(utils::fb_utilities::get_broadcast_address());
 266                     };
 267                     break;
 268                 case encrypt_what::rack:
 269                     so.filter_connection = [](const seastar::socket_address& addr) {
 270                         auto& snitch = locator::i_endpoint_snitch::get_local_snitch_ptr();
 271                         return snitch->get_datacenter(addr) == snitch->get_datacenter(utils::fb_utilities::get_broadcast_address())
 272                             && snitch->get_rack(addr) == snitch->get_rack(utils::fb_utilities::get_broadcast_address())
 273                             ;
 274                     };
 275                     break;
 276             }
```

Reply:
If we have " Allow gossip info lookup in production_snitch_base (and children) via "private" ip as well as public #9714 ", do we still need this patch?

Reply:
Encryption works because a.) Mostly people use the more sane "all" for what to encrypt, and b.) not all snitches actually use private networks afaict, especially not the ones we've used for testing. :-P

I think we need this, because it safeguards us against trying to downgrade in a situation where the lookup will not work (weird snitch?).

Reply:
Currently, EC2 + encrypt_what::rack/dc will no work, right? If so, I think this is a serious issue. 

Do you know any snich will still the lookup even with the fix of #9714? If yes, we need to fix them too.

Reply:
@elcallio please take a look at my example above: https://github.com/scylladb/scylla/pull/9709#discussion_r764511177

Reply:
Only `ec2_multi_region_snitch` and `gossiping_property_file_snitch` use/populate internal IP mappings to "preferred" IP:s in message service afaict. Thus only those two are subject to the problem and affected by this "fix". The problem only arises when we use actual non-endpoint addresses.